README.md
## Presentation
`mfetch` is an HTTP(S)-based file transfer program written in Golang, aimed at copying data between systems as fast
as possible over multiple TCP connections.  It can act as a client or a server, and has an interrupted transfer resuming
capability. In terms of performance, `mfetch` can easily saturate multi-gigabits network paths (even with back-to-back
latencies above 300ms), with a very moderate CPU usage (the main limitation being disk write iops).


## Usage
Invoking `mfetch` without argument (or with `-help`) will print the following documentation:
```
$ mfetch
usage:
  mfetch [<option...>] <argument...>

arguments:
  - client mode
        <source-url> [-|<local-file>|<target-url>]
  - server mode
        [<local-folder>]

options:
  -certificate string
        use provided TLS certificate & key in server mode (or "internal", no default)
  -chunksize int
        set worker chunksize (default 8388608)
  -concurrency int
        set transfer concurrency level (default 8)
  -dump
        dump HTTP requests and responses (default false)
  -insecure
        ignore remote TLS certificate errors (default false)
  -listen string
        set listening address & port in server mode (default client mode)
  -noresume
        disable transfer auto-resuming (default false)
  -password string
        set security password in server mode (no default)
  -post
        use HTTP POST method for remote target (default PUT)
  -progress
        emit transfer progress JSON indications (default false)
  -retries int
        set source request retries (default 4)
  -source value
        add HTTP header to source request (repeatable, no default)
  -target value
        add HTTP header to target request (repeatable, no default)
  -timeout int
        set requests timeout (default 10)
  -verbose
        set verbose mode (default false)
```


## Client mode
A valid `source-url` argument must be provided in client mode; if no target argument is provided, nothing will be
saved to disk (or written to remote target), but transfer statistics will still be printed on screen if the
`-verbose` option is provided, allowing to bench the considered network path before actually transferring documents.

The following options are available in client mode:

- `-chunksize` (default 8MiB): amount of memory used by each TCP connection (should not be changed, unless you
know what you are doing).

- `-concurrency` (default 8): number of concurrent TCP connections/HTTP requests (may be increased to maximize
transfer aggregated speed, as network latency between the client and server also increases).

- `-dump` (default false): dump requests and responses on standard error (mainly for debugging purpose).

- `-insecure` (default false): ignore invalid server TLS certificate (needed when using a self-signed server
certificate, like the `internal` one provided by `mfetch`, see `-certificate` below).

- `-noresume` (default false): always restart transfer from the beginning. <ins>Note</ins>: if the server does
not support byte-range requests, `concurrency` is automatically set to 1 and transfer resuming is disabled.

- `-post` (default PUT): use POST method (instead of PUT) in the remote `target-url` request.

- `-progress` (default false): emit transfer progress indications on standard output (in JSON format, see format
in the `Examples` section below).

- `-retries` (default 4): retry failed transfers for the specified number of times.

- `-source` (no default): additionnal HTTP headers sent with all source requests; can be used multiple times if
needed, for instance:
```
$ mfetch -source 'X-Header: value1' -source 'X-Another-Header: value2' https://...
```
- `-target` (no default): additionnal HTTP headers sent with the target request; can be used multiple times if
needed, for instance:
```
$ mfetch -target 'X-Header: value1' -target 'X-Another-Header: value2' https://...
```
- `-timeout` (default 10s): fail a connection if data transfer stay stalled for the specified number of seconds
(also see `retries` above).

- `-verbose` (default false): display transfer progress information on standard error (see format in the `Examples`
section below).


## Server mode
A `local-folder` argument may be provided in server mode, in which case only files in the specified folder will be
made accessible from an HTTP client (deeper folders won't be accessible). Files starting with `.` won't be accessible
either. If `mfetch` is started with no argument, it will server virtual files with sizes based on their names, for
benchmarking purpose (see syntax in the `Examples` section below). HTTP/2 is intentionally disabled (in HTTPS mode)
to make sure connecting clients use as many separate TCP connections as possible.

The following options are available in server mode:

- `-listen` (no default): activate `mfetch` server mode by specifying the (optional) IP address and TCP port to
listen to, for instance:
```
$ mfetch -listen 1.2.3.4:54321 ...
```
- `-certificate` (no default): switch the server to HTTPS (highly recommended if exposed to the public Internet);
either the string `"internal"` (in which case a self-signed [internal TLS certificate](server.go#L24-L29) is used),
or a comma-separated pair of files (certificate & key PEMs), for instance:
```
$ mfetch -listen ... -certificate /etc/ssl/certs/server-cert.pem,/etc/ssl/private/server-key.pem ...
```
- `-password` (no default): activate HTTP basic-authentication for all incoming requests (highly recommended if
the server is exposed to the public Internet).


## Examples
Starts an `mfetch` instance in "virtual files" server mode; clients requests matching the `/\d+[KMG]?i?B?` regex
pattern (for instance `/10M`, `/3GiB` or `/654321`) will be honoured by serving an all-zeroed content of the
corresponding size:
```
$ mfetch -listen :8000
```

Starts an `mfetch` instance in server mode and share the files in the /tmp folder, with HTTPS and password protection
activated (you may alternatively use an existing HTTP(S) server if you already have one handy):
```
$ mfetch -listen :443 -certificate internal -password password /tmp
```

Start an mfetch instance in client mode and download the `20GiB` file from the server instance above (since the
self-signed internal TLS certificate was used, the `-insecure` must be added to the command-line options) :
```
$ mfetch -verbose -progress -insecure https://login:password@localhost/20GiB out
860.45MB/20.0GB | 4.2% | 293.8Mb/s | 0:00:24/0:09:44
{"event":"start","size":21474836480,"received":0,"progress":0,"elapsed":0.000}
...
{"event":"progress","size":21474836480,"received":902247219,"progress":4,"elapsed":24.287}
```
When the `-verbose` option is specified on the command-line, `mfetch` will output some progress information
on the standard error in the following format:
```
<transferred size>/<total size> | <transferred percentage> | <transfer speed> | <elapsed time>/<total time>
```

When the `-progress` option is specified on the command-line, `mfetch` will emit some progress indications
on the standard output in the following JSON format:
```
{"event":"start|progress|end","size":<total size>,"received":<transferred size>,"progress":<transferred percentage>,"elapsed":<elapsed seconds>}
```

## Build and packaging
You need to install a recent version of the [Golang](https://golang.org/dl/) compiler (>= 1.17) and the GNU [make](https://www.gnu.org/software/make)
utility to build the `mfetch` binary. Once these requirements are fulfilled, clone the `mfetch` Github repository locally:
```
$ git clone https://github.com/pyke369/mfetch
```
and type:
```
$ make
```
This will take care of building everything. You may optionally produce a Debian binary package by typing:
```
$ make deb
```
(the [devscripts](https://packages.debian.org/fr/sid/devscripts) package needs to be installed for this last command to work)


## Projects with similar goals
- Facebook [WDT](https://github.com/facebook/wdt)


## License
MIT - Copyright (c) 2020-2021 Pierre-Yves Kerembellec


debian/changelog
mfetch (1.1.1) stable; urgency=medium

  * better let OS-es manage sockets buffers (esp. in LFN cases)

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Sun, 30 Jan 2022 19:15:53 +0000

mfetch (1.1.0) stable; urgency=medium

  * revamp the transfer algorithm and add remote target support

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Thu, 09 Dec 2021 10:37:15 +0100

mfetch (1.0.2) stable; urgency=medium

  * add JSON progress indications on stdout with the --progress option

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Tue, 06 Jul 2021 17:45:38 +0200

mfetch (1.0.1) stable; urgency=medium

  * initial public release

 -- Pierre-Yves Kerembellec <py.kerembellec@gmail.com>  Thu, 21 Jan 2021 10:17:43 +0100


util.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// internal self-signed X509 TLS certificate
// Serial:     58:c4:a8:c1:27:45:86:c0:e1:52:16:ab:d3:e4:3d:72:fb:b5:86:09
// Issuer:     CN = mfetch
// Subject:    CN = mfetch
// Not Before: Jan 21 15:04:26 2021 GMT
// Not After : Jan 19 15:04:26 2031 GMT
var tlscert, tlskey = `-----BEGIN CERTIFICATE-----
MIIDAzCCAeugAwIBAgIUWMSowSdFhsDhUhar0+Q9cvu1hgkwDQYJKoZIhvcNAQEL
BQAwETEPMA0GA1UEAwwGbWZldGNoMB4XDTIxMDEyMTE1MDQyNloXDTMxMDExOTE1
MDQyNlowETEPMA0GA1UEAwwGbWZldGNoMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8A
MIIBCgKCAQEA5fhnLwnmclzmHr01pobLlJ35UJkj1NJ3jWUVNkTnlITEBlkk9/Ta
35+VU8bLIBxgE9oKQyN0E5YGiT9D8uBp2jn7HAu8vDUFh1Ao8XEzdSgFjPtvPmH1
fSQ9b4qomDUDIebX8WDOox4VEwPcgz0Q0DF6cTmMs99kuyuhEyehdDRgSN/b7Qnf
FDAEJzXh2xZXTMjngPRF2cAHgGD6rh137JjUHoJ5kZ8iPxOp2cjaaw51HxafuFbN
aLb86n7g84hj51I8TzQjEM66rB1QvwsJVgBvSnX8ZkI+v4K5Y4ttzfQU2DD6wd9N
lIcdVPPmAtn5wPjMJyuC1HIAx8dNe/F+wwIDAQABo1MwUTAdBgNVHQ4EFgQUrj8O
XiNQfr1SSwtd6DlC/3Irb10wHwYDVR0jBBgwFoAUrj8OXiNQfr1SSwtd6DlC/3Ir
b10wDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOCAQEAQf1urVUoiQqT
6t0OZfeqYOLqahR+0l7Sm7TsDwM8u86WHpwXvbK4ULVX2FFtz0LI0bcVLbK4OxX/
4vNYMf3Q8ssjEKVyOKa3yy75/b7z5ahjMBqfWTETnSschJE+tuG5Nl4oGYwEYZXP
r6Ay1QTVsKlKLTG0+yiHbNPBVxNbXMSZEd4YBeVqikNmvy2WgVs6FhcWFMlrM5XO
eEFr2Z6t93jLzyYKyjxxomuijOCYy/oYNvbLisnbBb2OqwsjgsrMb5q324zXdBoD
7Ksvi+Ns0emN358FA210ORTxMMG5MSBY8OB7hphtxxi3slMzMsQ4FBEl5kKj9eFl
7DuHwqRfCA==
-----END CERTIFICATE-----
`,
	`-----BEGIN PRIVATE KEY-----
MIIEvwIBADANBgkqhkiG9w0BAQEFAASCBKkwggSlAgEAAoIBAQDl+GcvCeZyXOYe
vTWmhsuUnflQmSPU0neNZRU2ROeUhMQGWST39Nrfn5VTxssgHGAT2gpDI3QTlgaJ
P0Py4GnaOfscC7y8NQWHUCjxcTN1KAWM+28+YfV9JD1viqiYNQMh5tfxYM6jHhUT
A9yDPRDQMXpxOYyz32S7K6ETJ6F0NGBI39vtCd8UMAQnNeHbFldMyOeA9EXZwAeA
YPquHXfsmNQegnmRnyI/E6nZyNprDnUfFp+4Vs1otvzqfuDziGPnUjxPNCMQzrqs
HVC/CwlWAG9KdfxmQj6/grlji23N9BTYMPrB302Uhx1U8+YC2fnA+MwnK4LUcgDH
x0178X7DAgMBAAECggEAY0GJV3YQbn/GGrJTe6JmL6jXOIBARNTqIK7mLtwij6mV
6Z+EIzkdVrNMAjKk7SESHr9W+o9MxD9WZtpVe3h8d2HbDcnLFfhUgIiKg1r2eLRj
YOwMoYIqMG75zTCtf7Qxu+okfdvok+Kh+ekKveIXZaRVUpUiM2hR068LAHd0afB3
wE40boQ8kiRE31LLEqOV+Pzc/UWli7kH4ayhQLGhAIVeezGj+8E8FB4bU2GcvlEj
VZcB8Cxm42MSw9ucx/h+9DpG6RakgaBBD4lT9yxWt9oYOWX6AbvDSzn3jV/ueEct
Q5ho9WNqWvPXkOjbSk2aBge0cy5qipxY7f6quzN48QKBgQD1QOQ3Rbx3qZOtSSjC
fdT2RFrUCEQbMt7nXVgODQTqXBMTsaw7KjgurmCCroZ1pG0eI/g8eN06tktza8w8
tgiQq6GqrR0SBdZvaaIuVtm4yXetLLtSlhSJji/fIqnuqx/V2ygh2+2rNDEyo/aZ
708xRMltF7lwQiZ4YOUDHbsv/wKBgQDwDBL8kA4Kjkei6aOJbBLx438p18QbN7Lq
7qDaso4jgqwGnXRhSiiF7zPa3EuIj9cQ8TbmhcaCki7OtDyVChA1Ad1SaafgedSh
DzvpqAgQV/ZJIoGGP2kD0E9VfWWGhPBTF2kQ9VOB/ZgEK6epDL8JivUw4IncCt/t
dKmOUdjxPQKBgQDWRY2eJNVWjtexLBvqYNmxF2NroJUwVi+dYFZQYFuNDki0iiR5
xJc1YbB8PFLJcZDMJoz4+HgAlcgx3VqhKEEvdGRYo8qkNml1CYtihQrPgWWH7W7z
5p+m1o1InBZvqR61TzYu7uElFQJuxgXr08MSvpBlObcQNxs5TR6IrG8grQKBgQDg
UMglT5BveMmkiWQS9PU3KPoZ5dESBhihxWB3PcfpkyCiBd1NVPlNP1xbtuS2toOp
B1/gRz5bobMv4emC9KZ0gkuJycXg1LhH0W6RSD5Q14IEkcQr6XF+6NhZ8RZAgFX7
r7K08CubG5lEvG6uYITcrAe4JvtsrpTW1t/jaMSrmQKBgQCRcJDT1gJKnKpDKNKS
IaHIWZtLylFojEbmGrIS/5Xd8ez8EROhW5n+SZcenxT+2yVjJwhZAoELIXoPg09l
MJYHjCl6QyJlC4jv/NkchyPwYY11YubYXvxAjNnTB11uFLz2XpQF3+t5z60MBye/
GooiNcKqSQG5n1ivW1PdzY3T0Q==
-----END PRIVATE KEY-----
`

// repeated command-line option handler
type multiflag [][2]string

func (m *multiflag) String() string {
	return ""
}
func (m *multiflag) Set(value string) error {
	parts := strings.Split(value, ":")
	if len(parts) >= 2 {
		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(strings.Join(parts[1:], ":"))
		if len(key) > 0 && len(value) > 0 {
			*m = append(*m, [2]string{key, value})
		}
	}
	return nil
}

// duration human-display
func hduration(duration int64) string {
	if duration < 0 {
		return "-:--:--"
	}
	hours := duration / 3600
	duration -= (hours * 3600)
	minutes := duration / 60
	duration -= (minutes * 60)
	return fmt.Sprintf("%d:%02d:%02d", hours, minutes, duration)
}

// size human-display
func hsize(size int64) string {
	if size < (1 << 10) {
		return fmt.Sprintf("%dB", size)
	} else if size < (1 << 20) {
		return fmt.Sprintf("%.2fkB", float64(size)/(1<<10))
	} else if size < (1 << 30) {
		return fmt.Sprintf("%.2fMB", float64(size)/(1<<20))
	} else {
		return fmt.Sprintf("%.1fGB", float64(size)/(1<<30))
	}
}

// bandwidth human-display
func hbandwidth(bandwidth float64) string {
	if bandwidth < 1000 {
		return fmt.Sprintf("%.0fb/s", bandwidth)
	} else if bandwidth < 1000000 {
		return fmt.Sprintf("%.0fkb/s", bandwidth/1000)
	} else if bandwidth < 1000000000 {
		return fmt.Sprintf("%.1fMb/s", bandwidth/1000000)
	} else {
		return fmt.Sprintf("%.1fGb/s", bandwidth/1000000000)
	}
}

// error human-display & abort
func hfatal(message string, exit int) {
	if exit != 0 {
		fmt.Fprintf(os.Stderr, "\r                                                  \r%s - aborting\n", message)
	}
	if progress {
		fmt.Printf(`{"event":"error","message":"%s"}`+"\n", message)
	}
	os.Exit(exit)
}


client.go
package main

import (
	"context"
	"crypto/tls"
	"fmt"
	"io"
	"math"
	"net"
	"net/http"
	"net/http/httputil"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/pyke369/golang-support/bslab"
	"github.com/pyke369/golang-support/rcache"
)

type CHUNK struct {
	size  int
	start int
	end   int
	data  []byte
}

var (
	transport *http.Transport
	client    *http.Client
	buckets   = [11]int64{}
	cstart    = time.Now()
	istart    = 0
	ptype     = "start"
)

func request(input *CHUNK, dump bool) (output *CHUNK, err error) {
	ctx, cancel := context.WithCancel(context.Background())
	timer := time.AfterFunc(time.Duration(Timeout)*time.Second, func() {
		cancel()
	})
	defer timer.Stop()

	request, err := http.NewRequestWithContext(ctx, http.MethodGet, Flagset.Args()[0], nil)
	if err != nil {
		return input, err
	}
	request.Header.Set("User-Agent", fmt.Sprintf("%s/%s", PROGNAME, VERSION))
	for _, header := range Source {
		request.Header.Set(header[0], header[1])
	}
	request.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", input.start, input.end))
	if dump {
		headers, _ := httputil.DumpRequest(request, false)
		fmt.Fprintf(os.Stderr, "\r                                                       \r--- source request ---\n%s", headers)
	}

	response, err := client.Do(request)
	if err != nil {
		return input, err
	}
	if dump {
		headers, _ := httputil.DumpResponse(response, false)
		fmt.Fprintf(os.Stderr, "\r                                                       \r--- source response ---\n%s", headers)
	}
	captures := rcache.Get(`^bytes \d+-\d+/(\d+)$`).FindStringSubmatch(response.Header.Get("Content-Range"))
	if captures == nil || response.StatusCode != http.StatusPartialContent {
		response.Body.Close()
		return input, fmt.Errorf("source does not support partial requests")
	}
	input.size, _ = strconv.Atoi(captures[1])
	input.data = bslab.Get(input.end-input.start+1, nil)
	input.data = input.data[:cap(input.data)]
	read, offset := 0, 0
	for {
		read, err = response.Body.Read(input.data[offset:])
		timer.Reset(time.Duration(Timeout) * time.Second)
		if read > 0 {
			offset += read
		}
		if err == io.EOF {
			input.data, err = input.data[:offset], nil
			break
		}
		if err != nil {
			break
		}
	}
	response.Body.Close()
	if err == nil && (input.size <= 0 || len(input.data) != input.end-input.start+1) {
		err = fmt.Errorf("truncated source content")
	}
	if err != nil {
		bslab.Put(input.data)
	}
	return input, err
}

func hsize(size int) string {
	if size < (1 << 10) {
		return fmt.Sprintf("%dB", size)
	} else if size < (1 << 20) {
		return fmt.Sprintf("%.2fkiB", float64(size)/(1<<10))
	} else if size < (1 << 30) {
		return fmt.Sprintf("%.2fMiB", float64(size)/(1<<20))
	} else {
		return fmt.Sprintf("%.1fGiB", float64(size)/(1<<30))
	}
}
func hduration(duration int) string {
	if duration < 0 {
		return "-:--:--"
	}
	hours := duration / 3600
	duration -= (hours * 3600)
	minutes := duration / 60
	duration -= (minutes * 60)
	return fmt.Sprintf("%d:%02d:%02d", hours, minutes, duration)
}
func hbandwidth(bandwidth float64) string {
	if bandwidth < 1000 {
		return fmt.Sprintf("%.0fb/s", bandwidth)
	} else if bandwidth < (1000 * 1000) {
		return fmt.Sprintf("%.0fkb/s", bandwidth/(1000))
	} else if bandwidth < (1000 * 1000 * 1000) {
		return fmt.Sprintf("%.1fMb/s", bandwidth/(1000*1000))
	} else {
		return fmt.Sprintf("%.1fGb/s", bandwidth/(1000*1000*1000))
	}
}
func stats(start, size int, end bool) {
	for index := 10; index >= 1; index-- {
		atomic.StoreInt64(&buckets[index], atomic.LoadInt64(&buckets[index-1]))
	}
	atomic.AddInt64(&buckets[0], -atomic.LoadInt64(&buckets[1]))
	total, divider := 0, 0
	for index := 1; index <= 10; index++ {
		if buckets[index] != 0 {
			total += int(buckets[index])
			divider++
		}
	}
	if Verbose {
		bandwidth, duration, elapsed := float64(0), -1, int(time.Now().Sub(cstart)/time.Second)
		if divider != 0 {
			bandwidth = (float64(total) * 8) / (float64(divider) / 5)
			duration = int((float64(size-istart) * 8) / bandwidth)
			if duration < elapsed {
				duration = elapsed
			}
		}
		fmt.Fprintf(os.Stderr, "\r%s/%s | %.1f%% | %s | %s/%s     ", hsize(start), hsize(size),
			(float64(start)*100)/float64(size), hbandwidth(bandwidth), hduration(elapsed), hduration(duration))
		if end {
			fmt.Fprintf(os.Stderr, "\n")
		}
	}
	if Progress && ptype != "end" {
		if end {
			ptype = "end"
		}
		fmt.Printf(`{"event":"%s","size":%d,"received":%d,"progress":%.1f,"elapsed":%.3f}`+"\n",
			ptype, size, start, (float64(start)*100)/float64(size), float64(time.Now().Sub(cstart))/float64(time.Second))
		if ptype == "start" {
			ptype = "progress"
		}
	}
}

func exit(message interface{}, exit int) {
	if exit != 0 {
		fmt.Fprintf(os.Stderr, "\r                                                       \r%v - aborting\n", message)
	}
	if Progress {
		fmt.Printf(`{"event":"error","message":"%s"}`+"\n", message)
	}
	os.Exit(exit)
}

func Client() {
	var (
		file   *os.File
		reader *io.PipeReader
		writer *io.PipeWriter
	)

	if Flagset.NArg() < 1 {
		Flagset.Usage()
		os.Exit(1)
	}
	transport = &http.Transport{
		DialContext:           (&net.Dialer{Timeout: time.Duration(Timeout) * time.Second}).DialContext,
		TLSClientConfig:       &tls.Config{InsecureSkipVerify: Insecure},
		TLSHandshakeTimeout:   time.Duration(Timeout) * time.Second,
		ResponseHeaderTimeout: time.Duration(Timeout) * time.Second,
		ReadBufferSize:        1 << 20,
		MaxIdleConnsPerHost:   32,
	}
	client = &http.Client{Transport: transport}

	source, err := request(&CHUNK{}, Dump)
	if err != nil {
		exit(err, 1)
	}
	target, start := "", 0
	if Flagset.NArg() > 1 {
		target = Flagset.Args()[1]
		if target != "-" && !strings.HasPrefix(target, "http") && !Noresume {
			if info, err := os.Stat(target); err == nil && info.Mode().IsRegular() && int(info.Size()) <= source.size {
				start = int(info.Size())
			}
		}
		if target == "-" {
			Progress = false
		}
	}
	istart = start

	queue, offset := make(chan *CHUNK, 32), start
	go func() {
		for offset < source.size {
			waiter, concurrency := &sync.WaitGroup{}, int(math.Max(1, math.Min(float64((source.size-offset)/Chunksize), float64(Concurrency))))
			for index := 0; index < concurrency; index++ {
				waiter.Add(1)
				go func(start, end int) {
					tries := 0
					for tries < Retries {
						if chunk, err := request(&CHUNK{start: start, end: end}, false); err == nil {
							queue <- chunk
							waiter.Done()
							return
						}
						tries++
					}
					queue <- &CHUNK{}
				}(offset+(Chunksize*index), int(math.Min(float64(source.size-1), float64(offset+(Chunksize*(index+1))-1))))
			}
			waiter.Wait()
			offset += Chunksize * concurrency
			offset = int(math.Min(float64(source.size), float64(offset)))
		}
		close(queue)
	}()

	go func() {
		stats(start, source.size, false)
		for range time.Tick(time.Second / 5) {
			stats(start, source.size, false)
			if start >= source.size {
				break
			}
		}
	}()

	chunks, done := map[int]*CHUNK{}, make(chan struct{})
	for start < source.size {
		chunk := <-queue
		if chunk.data == nil {
			exit("truncated source content", 2)
		}
		chunks[chunk.start] = chunk
		for {
			found := false
			for key, chunk := range chunks {
				if key == start {
					switch {
					case target == "":
					case target == "-":
						if _, err := os.Stdout.Write(chunk.data); err != nil {
							exit(err, 3)
						}
					case strings.HasPrefix(target, "http"):
						if writer == nil {
							method := http.MethodPut
							if Post {
								method = http.MethodPost
							}
							reader, writer = io.Pipe()
							request, err := http.NewRequest(method, target, reader)
							if err != nil {
								exit(err, 4)
							}
							request.Header.Set("User-Agent", fmt.Sprintf("%s/%s", PROGNAME, VERSION))
							request.Header.Set("Content-Type", "application/octet-stream")
							for _, header := range Target {
								request.Header.Set(header[0], header[1])
							}
							request.ContentLength = int64(source.size)
							if Dump {
								headers, _ := httputil.DumpRequest(request, false)
								fmt.Fprintf(os.Stderr, "\r                                                       \r--- target request ---\n%s", headers)
							}
							go func() {
								response, err := client.Do(request)
								if err != nil {
									exit(err, 4)
								}
								if Dump {
									headers, _ := httputil.DumpResponse(response, false)
									fmt.Fprintf(os.Stderr, "\r                                                       \r--- target response ---\n%s", headers)
								}
								response.Body.Close()
								if response.StatusCode/100 != 2 {
									exit(fmt.Sprintf("invalid target status code %d", response.StatusCode), 4)
								}
								close(done)
							}()
						}
						writer.Write(chunk.data)

					default:
						if file == nil {
							os.MkdirAll(filepath.Dir(target), 0755)
							value, err := os.OpenFile(target, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
							if err != nil {
								exit(err, 5)
							}
							file = value
							file.Truncate(int64(start))
						}
						if _, err := file.Write(chunk.data); err != nil {
							exit(err, 5)
						}
					}
					atomic.AddInt64(&buckets[0], int64(len(chunk.data)))
					bslab.Put(chunk.data)
					start = chunk.end + 1
					delete(chunks, key)
					found = true
					break
				}
			}
			if !found {
				break
			}
		}
	}
	if file != nil {
		file.Close()
	}
	if writer != nil {
		<-done
		writer.Close()
	}
	stats(start, source.size, true)
}


main.go
package main

import (
	"flag"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strings"

	"github.com/pyke369/golang-support/multiflag"
)

var (
	PROGNAME = "mfetch"
	VERSION  = "1.1.1"

	Flagset     = flag.NewFlagSet(PROGNAME, flag.ExitOnError)
	Concurrency = 8
	Chunksize   = 8 << 20
	Timeout     = 10
	Retries     = 4
	Source      = multiflag.Multiflag{}
	Target      = multiflag.Multiflag{}
	Post        = false
	Insecure    = false
	Noresume    = false
	Verbose     = false
	Dump        = false
	Progress    = false
	Listen      = ""
	Certificate = ""
	Password    = ""
)

func main() {
	Flagset.Usage = func() {
		fmt.Fprintf(os.Stderr, `usage:
  %s [<option...>] <argument...>

arguments:
  - client mode
        <source-url> [-|<local-file>|<target-url>]
  - server mode
        [<local-folder>]

options:
`, filepath.Base(os.Args[0]))
		Flagset.PrintDefaults()
	}
	Flagset.IntVar(&Concurrency, "concurrency", Concurrency, "set transfer concurrency level")
	Flagset.IntVar(&Chunksize, "chunksize", Chunksize, "set worker chunksize")
	Flagset.IntVar(&Timeout, "timeout", Timeout, "set requests timeout")
	Flagset.IntVar(&Retries, "retries", Retries, "set source request retries")
	Flagset.Var(&Source, "source", "add HTTP header to source request (repeatable, no default)")
	Flagset.Var(&Target, "target", "add HTTP header to target request (repeatable, no default)")
	Flagset.BoolVar(&Post, "post", Post, "use HTTP POST method for remote target (default PUT)")
	Flagset.BoolVar(&Insecure, "insecure", Insecure, "ignore remote TLS certificate errors (default false)")
	Flagset.BoolVar(&Noresume, "noresume", Noresume, "disable transfer auto-resuming (default false)")
	Flagset.BoolVar(&Verbose, "verbose", Verbose, "set verbose mode (default false)")
	Flagset.BoolVar(&Dump, "dump", Dump, "dump HTTP requests and responses (default false)")
	Flagset.BoolVar(&Progress, "progress", Progress, "emit transfer progress JSON indications (default false)")
	Flagset.StringVar(&Listen, "listen", Listen, "set listening address & port in server mode (default client mode)")
	Flagset.StringVar(&Certificate, "certificate", Certificate, `use provided TLS certificate & key in server mode (or "internal", no default)`)
	Flagset.StringVar(&Password, "password", Password, "set security password in server mode (no default)")
	Flagset.Parse(os.Args[1:])
	Concurrency = int(math.Min(32, math.Max(1, float64(Concurrency))))
	Chunksize = int(math.Min(64<<20, math.Max(1<<20, float64(Chunksize))))
	Timeout = int(math.Min(30, math.Max(1, float64(Timeout))))
	Retries = int(math.Min(4, math.Max(0, float64(Retries))))
	Listen, Certificate, Password = strings.TrimLeft(strings.TrimSpace(Listen), "*"), strings.TrimSpace(Certificate), strings.TrimSpace(Password)
	if Listen == "" {
		Client()
	} else {
		Server()
	}
}
